sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/JacksonAdapter.java

```diff
@@ -203,28 +213,41 @@ public String serializeList(List<?> list, CollectionFormat format) {
         final JavaType javaType = createJavaType(type);
         try {
             if (encoding == SerializerEncoding.XML) {
-                return (T) xmlMapper.readValue(inputStream, javaType);
+                return xmlMapper.readValue(inputStream, javaType);
             } else {
-                return (T) serializer().readValue(inputStream, javaType);
+                return serializer().readValue(inputStream, javaType);
             }
         } catch (JsonParseException jpe) {
             throw logger.logExceptionAsError(new MalformedValueException(jpe.getMessage(), jpe));
         }
     }
 
-    @SuppressWarnings("deprecation")
     @Override
     public <T> T deserialize(HttpHeaders headers, Type deserializedHeadersType) throws IOException {
         if (deserializedHeadersType == null) {
             return null;
         }
 
-        final String headersJsonString = headerMapper.writeValueAsString(headers);
-        T deserializedHeaders =
-            headerMapper.readValue(headersJsonString, createJavaType(deserializedHeadersType));
+        T deserializedHeaders = headerMapper.convertValue(headers, createJavaType(deserializedHeadersType));
 
         final Class<?> deserializedHeadersClass = TypeUtil.getRawClass(deserializedHeadersType);
         final Field[] declaredFields = deserializedHeadersClass.getDeclaredFields();
+
+        /*
+         * A list containing all handlers for header collections of the header type.
+         */
+        final List<HeaderCollectionHandler> headerCollectionHandlers = new ArrayList<>();
+
+        /*
+         * This set is an optimization where we track the first character of all HeaderCollections defined on the
+         * deserialized headers type. This allows us to optimize away startWiths checks which are much more costly than
+         * getting the first character.
+         */
+        final Set<Character> headerCollectionsFirstCharacters = new HashSet<>();
```

Comment:
Not sure if this optimization is worth the complexity. `String`'s `startsWith` also fails fast if the first char doesn't match. Was there any significant gain from this optimization?

Reply:
I didn't know this, I can run some quick validations.

Reply:
The usage of this was in the wrong spot, this handles no-op'ing a header if it doesn't begin with a character that starts any header collection in the deserialized header type.